src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/MetadataReaderExtensions.NativeFormat.cs

```diff
@@ -79,20 +79,21 @@ internal static Type[] GetCustomModifiers(this Handle handle, MetadataReader rea
             if (handleType != HandleType.ModifiedType)
                 return Array.Empty<Type>();
 
-            LowLevelList<Type> customModifiers = new LowLevelList<Type>();
+            ArrayBuilder<Type> customModifiers = default;
             do
             {
                 NativeFormatModifiedType modifiedType = handle.ToModifiedTypeHandle(reader).GetModifiedType(reader);
                 if (optional == modifiedType.IsOptional)
                 {
                     Type customModifier = modifiedType.ModifierType.Resolve(reader, typeContext).ToType();
-                    customModifiers.Insert(0, customModifier);
+                    customModifiers.Add(customModifier);
                 }
 
                 handle = modifiedType.Type;
                 handleType = handle.HandleType;
             }
             while (handleType == HandleType.ModifiedType);
+            customModifiers.AsSpan(0).Reverse();
```

Comment:
When I added the `ArrayBuilder<T>.AsSpan(int start)` in https://github.com/dotnet/runtime/pull/102394, I did not limit the returned `Span<T>` to the only used portion, so I believe this this `Reverse()` is currently reversing the entire underlying `ArrayBuilder<T>._items` array 😅 

I think we need to update `ArrayBuilder<T>.AsSpan(int start)` to slice the only used portion in order for this to do the correct thing.

(This concern was already raised in https://github.com/dotnet/runtime/pull/102394/files#r1605898748)

Reply:
Digging further, it also came from the early era of CoreRT.
Now we have the more-optimized version of ArrayBuilder in CoreLib. We can borrow some optimizations there. I'm a little concerned about size optimization to directly use that version, especially calls to the monster `Array.Copy`.